File: architecture-skills/design-patterns/examples/command/command.php
<?php
// Command pattern in PHP.
//
// Text editor with undo: a Command interface with execute/undo, concrete
// commands capturing their own undo state, and a history stack as the
// invoker. For one-shot commands without undo, a closure is the lighter
// idiom (framework job queues are commands in production form).

declare(strict_types=1);

final class Editor // receiver
{
    public string $text = '';
}

interface Command
{
    public function execute(): void;

    public function undo(): void;
}

final class InsertText implements Command
{
    public function __construct(
        private readonly Editor $editor,
        private readonly string $text,
    ) {
    }

    public function execute(): void
    {
        $this->editor->text .= $this->text;
    }

    public function undo(): void
    {
        $this->editor->text = substr($this->editor->text, 0, -strlen($this->text));
    }
}

final class DeleteLast implements Command
{
    private string $deleted = ''; // captured on execute so undo can restore it

    public function __construct(
        private readonly Editor $editor,
        private readonly int $count,
    ) {
    }

    public function execute(): void
    {
        $this->deleted = substr($this->editor->text, -$this->count);
        $this->editor->text = substr($this->editor->text, 0, -$this->count);
    }

    public function undo(): void
    {
        $this->editor->text .= $this->deleted;
    }
}

final class History // invoker
{
    /** @var Command[] */
    private array $undoStack = [];

    public function execute(Command $command): void
    {
        $command->execute();
        $this->undoStack[] = $command;
    }

    public function undo(): void
    {
        $command = array_pop($this->undoStack);
        $command?->undo();
    }
}

$editor = new Editor();
$history = new History();

$history->execute(new InsertText($editor, 'Hello'));
$history->execute(new InsertText($editor, ', world!'));
$history->execute(new DeleteLast($editor, 1));
printf("after edits: \"%s\"\n", $editor->text);

$history->undo();
$history->undo();
printf("after 2 undos: \"%s\"\n", $editor->text);

File: architecture-skills/design-patterns/examples/command/command.cs
// Command pattern in C#.
//
// Text editor with undo: an ICommand-style interface with Execute/Undo,
// concrete commands capturing their own undo state, and a history stack
// as the invoker. For one-shot commands without undo, an Action delegate
// is the lighter idiom (WPF's ICommand is this pattern in the framework).

using System;
using System.Collections.Generic;
using System.Text;

public sealed class Editor // receiver
{
    public StringBuilder Text { get; } = new();
}

public interface ICommand
{
    void Execute();
    void Undo();
}

public sealed class InsertText : ICommand
{
    private readonly Editor _editor;
    private readonly string _text;

    public InsertText(Editor editor, string text)
    {
        _editor = editor;
        _text = text;
    }

    public void Execute() => _editor.Text.Append(_text);

    public void Undo() => _editor.Text.Length -= _text.Length;
}

public sealed class DeleteLast : ICommand
{
    private readonly Editor _editor;
    private readonly int _count;
    private string _deleted = ""; // captured on Execute so Undo can restore it

    public DeleteLast(Editor editor, int count)
    {
        _editor = editor;
        _count = count;
    }

    public void Execute()
    {
        int start = _editor.Text.Length - _count;
        _deleted = _editor.Text.ToString(start, _count);
        _editor.Text.Length = start;
    }

    public void Undo() => _editor.Text.Append(_deleted);
}

public sealed class History // invoker
{
    private readonly Stack<ICommand> _undoStack = new();

    public void Execute(ICommand command)
    {
        command.Execute();
        _undoStack.Push(command);
    }

    public void Undo()
    {
        if (_undoStack.Count > 0)
        {
            _undoStack.Pop().Undo();
        }
    }
}

public static class Program
{
    public static void Main()
    {
        var editor = new Editor();
        var history = new History();

        history.Execute(new InsertText(editor, "Hello"));
        history.Execute(new InsertText(editor, ", world!"));
        history.Execute(new DeleteLast(editor, 1));
        Console.WriteLine($"after edits: \"{editor.Text}\"");

        history.Undo();
        history.Undo();
        Console.WriteLine($"after 2 undos: \"{editor.Text}\"");
    }
}
